action/penalization/allegation.go

```diff
@@ -0,0 +1,157 @@
+
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, uint64(r.BlockHeight))
+	tag4 := kv.Pair{
+		Key:   []byte("tx.height"),
+		Value: b,
+	}
+	tag5 := kv.Pair{
+		Key:   []byte("tx.proof"),
+		Value: []byte(r.ProofMsg),
+	}
+	tag6 := kv.Pair{
+		Key:   []byte("tx.requestID"),
+		Value: []byte(r.RequestID),
+	}
+
+	tags = append(tags, tag, tag2, tag3, tag4, tag5, tag6)
+	return tags
+}
+
+var _ action.Tx = allegationTx{}
+
+type allegationTx struct{}
+
+func (atx allegationTx) Validate(ctx *action.Context, tx action.SignedTx) (bool, error) {
+	r := &Allegation{}
+	err := r.Unmarshal(tx.Data)
+	if err != nil {
+		return false, errors.Wrap(action.ErrWrongTxType, err.Error())
+	}
+
+	err = action.ValidateBasic(tx.RawBytes(), r.Signers(), tx.Signatures)
+	if err != nil {
+		return false, err
+	}
+
+	err = action.ValidateFee(ctx.FeePool.GetOpt(), tx.Fee)
+	if err != nil {
+		return false, err
+	}
+
+	if err := r.ValidatorAddress.Err(); err != nil {
+		return false, err
+	}
+
+	if err := r.MaliciousAddress.Err(); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (atx allegationTx) ProcessCheck(ctx *action.Context, tx action.RawTx) (ok bool, result action.Response) {
+	ctx.Logger.Debug("Processing 'allegation' transaction for ProcessCheck", tx)
+	ok, result = runAllegationTransaction(ctx, tx)
+	ctx.Logger.Debug("Result 'allegation' transaction for ProcessCheck", ok, result)
+	return
+}
+
+func (atx allegationTx) ProcessDeliver(ctx *action.Context, tx action.RawTx) (ok bool, result action.Response) {
+	ctx.Logger.Debug("Processing 'allegation' transaction for ProcessDeliver", tx)
+	ok, result = runAllegationTransaction(ctx, tx)
+	ctx.Logger.Debug("Result 'allegation' transaction for ProcessDeliver", ok, result)
+	return
+}
+
+func (atx allegationTx) ProcessFee(ctx *action.Context, signedTx action.SignedTx, start action.Gas, size action.Gas) (bool, action.Response) {
+	ctx.Logger.Debug("Processing 'allegation' Transaction for ProcessFee", signedTx)
+	r := &Allegation{}
+	err := r.Unmarshal(signedTx.Data)
+	if err != nil {
+		return false, action.Response{Log: errors.Wrap(err, "failed to unmarshal").Error()}
+	}
+	return action.StakingPayerFeeHandling(ctx, r.ValidatorAddress, signedTx, start, size, 1)
+}
+
+func runAllegationTransaction(ctx *action.Context, tx action.RawTx) (bool, action.Response) {
+	al := &Allegation{}
+	err := al.Unmarshal(tx.Data)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, action.ErrWrongTxType, al.Tags(), err)
+	}
+
+	if ctx.EvidenceStore.IsFrozenValidator(al.MaliciousAddress) {
+		return helpers.LogAndReturnFalse(ctx.Logger, evidence.ErrFrozenValidator, al.Tags(), err)
+	}
+
+	if !ctx.EvidenceStore.IsActiveValidator(al.ValidatorAddress) {
+		return helpers.LogAndReturnFalse(ctx.Logger, evidence.ErrNonActiveValidator, al.Tags(), err)
+	}
+
+	if al.ValidatorAddress.Equal(al.MaliciousAddress) {
+		return helpers.LogAndReturnFalse(ctx.Logger, action.ErrInvalidAddress, al.Tags(), err)
+	}
+
```

Comment:
We need to check , that we dont already have a allegation request against the malicious validator .

Test case 
Req 1 : Create Alegation  ( Validator 1  -> validator 2)
Req 2 : Create Alegation  ( Validator 1  -> validator 2)
Vote yes : req 1
Vote yes : req 2

Validator 2 gets two deductions in Stake ( Should not happen)

Reply:
it covered by IsRequestIDBusy function, see the code

Reply:
We are talking about two different user IDs here , the same validator cannot have do multiple alegations

Reply:
It was not in functional requirements and never discussed at weekly call, I guess this exception was not realized in code. Please clarify with workflow.

Reply:
I think this is an implementation detail ,and not a a functional requirement @pav-kap ?